[nnc] Add support for dynamic shapes in TensorExprKernel

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67197

Differential Revision: [D31902471](https://our.internmc.facebook.com/intern/diff/D31902471)

navahgar committed:
[nnc] Add support for dynamic shapes in TensorExprKernel

[ghstack-poisoned]

navahgar committed:
Update on "[nnc] Add support for dynamic shapes in TensorExprKernel"


Differential Revision: [D31902471](https://our.internmc.facebook.com/intern/diff/D31902471)

[ghstack-poisoned]

eellison reviewed (approved):
Sweet, looks good to me ! Accepting but it might be good to get one more review from @bertmaher or @ZolotukhinM

navahgar committed:
Update on "[nnc] Add support for dynamic shapes in TensorExprKernel"


Differential Revision: [D31902471](https://our.internmc.facebook.com/intern/diff/D31902471)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -820,16 +835,61 @@ static std::vector<ExprHandle> toExprHandles(const std::vector<T>& sizes) {
+      strides.push_back(LongImm::make(cs));
+    }
+  } else {
+    throw std::runtime_error("Only concrete strides are handled");
```
Nit: why not an assertion or TE error?

navahgar replied:
I didn't have a preference either way. Changed it to an assertion.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -209,6 +211,12 @@ class TORCH_API TensorExprKernel {
   // push the buf args into the stack so NNC IR can access them at runtime.
   void preAllocIntermediateBufs(std::unordered_set<BufPtr>& interm_bufs);
 
+  ExprHandle getVarForShape(const c10::ShapeSymbol& ss);
+  std::vector<ExprHandle>& getStridesForValue(const Value* v);
+  BufHandle bindSymbolicShapeInput(const Value* input, const std::string& name);
+  std::vector<ExprHandle> sizesFromSymbolicShape(
+      const c10::SymbolicShape& shape);
```
Can we make these methods private?

navahgar replied:
Done

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1234,6 +1353,30 @@ std::vector<CodeGen::CallArg> TensorExprKernel::prepareRunArgs(
+    // symbolic shape input params passed in to this call.
+    TORCH_INTERNAL_ASSERT(
+        tensorOutputSymbolicSizes_.size() == bufOutputs_.size());
+    tensorOutputSizes_.resize(bufOutputs_.size());
```
Why do we need to resize it on every invocation of the kernel? Shouldn't it have the appropriate size already?

navahgar replied:
Only the first `resize` should have any effect. The subsequent ones should not do anything. 

I have now moved the `resize` to earlier, during compilation and made these asserts.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -95,17 +95,20 @@ class TORCH_API TensorExprKernel {
       const std::shared_ptr<Graph>& subgraph,
       std::unordered_map<c10::Symbol, NNCLoweringFunction> custom_lowerings =
           {},
+      std::vector<int64_t> symbolic_shape_inputs = {},
```
I think we need to start adding documentation about all the arguments. We keep piling them on (not your fault :) ), and most of them are used for completely different things - it might be difficult to grasp this code for new people.

navahgar replied:
I added a brief description of the arguments above. We can update it with more details as needed.

ZolotukhinM reviewed (approved):
Overall looks good to me!

navahgar committed:
Update on "[nnc] Add support for dynamic shapes in TensorExprKernel"


Differential Revision: [D31902471](https://our.internmc.facebook.com/intern/diff/D31902471)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31902471).

navahgar committed:
Update on "[nnc] Add support for dynamic shapes in TensorExprKernel"


Differential Revision: [D31902471](https://our.internmc.facebook.com/intern/diff/D31902471)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31902471).

navahgar committed:
Update on "[nnc] Add support for dynamic shapes in TensorExprKernel"


Differential Revision: [D31902471](https://our.internmc.facebook.com/intern/diff/D31902471)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31902471).

ngimel commented:
Reverting, breaks master CI, see e.g. https://github.com/pytorch/pytorch/runs/4097495799?check_suite_focus=true
